stylix/palette.nix

```
@@ -168,23 +165,20 @@ in
     # This attrset can be used like a function too, see
     # https://github.com/SenchoPens/base16.nix/blob/b390e87cd404e65ab4d786666351f1292e89162a/README.md#theme-step-22
     #
-    # NOTE: base16Scheme is nullable. If null, we cannot pass it to mkSchemeAttrs.
-    # See https://github.com/nix-community/stylix/pull/1408
+    # NOTE: base16Scheme doesn't always have a default definition.
+    #
+    # We add an assertion here, because this value is used throughout Stylix.
+    # Using the assertions option is ineffective because it is checked after
+    # the errors caused by reading `base16Scheme` have occurred.
     lib.stylix.colors =
-      if cfg.base16Scheme == null then
-        null
-      else
-        (cfg.base16.mkSchemeAttrs cfg.base16Scheme).override cfg.override;
-
-    assertions = lib.mkIf cfg.enable [
-      {
-        assertion = !(cfg.image == null && cfg.base16Scheme == null);
-
-        message = ''
-          stylix: one of `stylix.image` or `stylix.base16Scheme` must be set
-        '';
-      }
-    ];
+      lib.throwIfNot opts.base16Scheme.isDefined
+        "stylix: one of `${opts.image}' or `${opts.base16Scheme}' must be set."
+        ((cfg.base16.mkSchemeAttrs cfg.base16Scheme).override cfg.override);
+
+    # Set the default here so it can be conditional
+    stylix.base16Scheme = lib.mkIf (cfg.image != null) (
+      lib.mkOptionDefault cfg.generated.palette
+    );
```

It may be nicer to move the assertion here actually. That way users will never see a "used but not defined error"...


```suggestion
    # Set the default here so it can be conditional
    stylix.base16Scheme = lib.mkOptionDefault (
      if cfg.image == null then
        throw "stylix: ........"
      else
        cfg.generated.palette
    );
```

or

```suggestion
    # Set the default here so it can be conditional
    stylix.base16Scheme = lib.mkOptionDefault (
      lib.throwIfNot (cfg.image != null)
        "stylix: ........"
        cfg.generated.palette
    );
```